File: moodle/scope_hysteresis_area.py
"""Python translation of scope_hysteresis_area.m."""
import numpy as np
import matplotlib.pyplot as plt


def scope_hysteresis_area(file_name, delta):
    a = np.loadtxt(file_name)
    a[:, 0] -= a[:, 0].min()
    a[:, 1] -= a[:, 1].min()
    x, y = a[:, 0], a[:, 1]
    x_min, x_max = x.min(), x.max()
    y_min, y_max = y.min(), y.max()

    plt.plot(x, y)
    plt.grid(True)

    dx = delta
    N = int(np.ceil((x_max - x_min) / dx))
    new_x, upper, lower = [], [], []
    for i in range(1, N + 1):
        left = x_min + (i - 1) * dx
        right = x_min + i * dx
        mask = (x >= left) & (x < right)
        if not mask.any():
            raise ValueError("ERROR: The steps are too small")
        y_bin = y[mask]
        hi, lo = y_bin.max(), y_bin.min()
        edge = i <= 3 or i > N - 3
        if edge or abs(hi - lo) > 0.1:
            new_x.append((left + right) / 2)
            upper.append(hi)
            lower.append(lo)

    new_x = np.array(new_x)
    upper = np.array(upper)
    lower = np.array(lower)

    plt.plot(new_x, upper, 'r')
    plt.plot(new_x, lower, 'g')

    max_sec_derv = max(np.diff(lower, n=2).max(), np.diff(upper, n=2).max())
    area_error = max_sec_derv * (x_max - x_min) ** 3 / N ** 2
    area = np.trapezoid(upper, new_x) - np.trapezoid(lower, new_x)

    plt.text(x_min, (y_max + y_min) / 2,
             f'Area = {area} \\pm {area_error}', fontsize=15)
    plt.show()
    return area, area_error


if __name__ == '__main__':
    import sys
    scope_hysteresis_area(sys.argv[1], float(sys.argv[2]))

File: ferromagnetism/moodle/scope_hysteresis_area.m
%clear all;
%close all;
function[] = scope_hysteresis_area(file_name,delta)
close all;
%a=load('scope_output.txt');
a=load(file_name);
%a(:,2)= a(:,2);
x_min = min(a(:,1));
y_min=min(a(:,2));
a(:,2)=a(:,2)-y_min;
a(:,1)=a(:,1)-x_min;
x_max = max(a(:,1));
y_max=max(a(:,2));
plot(a(:,1),a(:,2));
grid on;
hold on;
x_max = max(a(:,1));
x_min = min(a(:,1));
y_min=min(a(:,2));
%dx=0.1;
dx=delta;
N=ceil( (x_max-x_min)/dx)  ;
 %tmp_x_array(1)=10;
 index=1;
for i=1:N
    k=1;
    for j=1:size(a(:,1))
         if  ( (x_min+(i-1)*dx <=  a(j,1) )  & (  a(j,1) < x_min+i*dx ) )
                 
                  tmp_x_array(k)=a(j,2);             
                   k=k+1;
         end
              
    end

            
            upper_loop_max = max(tmp_x_array);
            lower_loop_min =min(tmp_x_array);
         if(size(upper_loop_max)~=1 | size(lower_loop_min)~=1  )
             error('ERROR: The steps are too small '); 
           
         end
            
            
            if(abs(upper_loop_max-lower_loop_min)>0.1 & (i < N-3)  & (  i > 3)   ) 
                new_x(index) = (x_min+i*dx + x_min+(i-1)*dx )/2;
                upper_loop(index) = upper_loop_max;
                lower_loop(index) = lower_loop_min;
                index=index+1;
                
            end
            
            if((i > N-3 )   | (  i < 3 )   )
                new_x(index) = (x_min+i*dx + x_min+(i-1)*dx )/2;
                upper_loop(index) = upper_loop_max;
                lower_loop(index) = lower_loop_min;
                index=index+1;
                
                
            end
          tmp_x_array(:)=[];
 end
plot(new_x,upper_loop,'r');
hold on;
plot(new_x,lower_loop,'g');


max_sec_derv = max(max(diff(lower_loop,2) ) ,max(diff(upper_loop,2)) ) ;
area_error = max_sec_derv*(x_max-x_min)^3 /(N*N);
area = trapz(new_x,upper_loop) - trapz(new_x,lower_loop);

str_area = num2str(area);
str_error = num2str(area_error);
Str = strcat('Area =', str_area,'\pm', str_error);

text((x_min),(y_max+y_min)/2,Str  ,'FontSize',15);


